File: challenges/2020/07-handyHaversacks/python/partTwo.py
from common import *
from typing import Dict


def count_for_bag(ruleset: Dict[str, Dict[str, int]], test_cl: str) -> int:
    if len(ruleset[test_cl]) == 0:
        return -1

    count = 0

    for child_colour in ruleset[test_cl]:
        child_bags = count_for_bag(ruleset, child_colour)
        if child_bags == -1:
            v = ruleset[test_cl][child_colour]
        else:
            v = ruleset[test_cl][child_colour] * child_bags
            # The below line includes the number of bags that contain all the children - for
            # example, if you had a bag type that had 6 child bags, and there were 3 of these
            # master bags, the line above would add all the child bags (6*3 of them), and the line
            # below would add the 3 container bags.
            v += ruleset[test_cl][child_colour]
        count += v

    return count


def partTwo(instr: str) -> int:
    rules = parse(instr)
    return count_for_bag(rules, target_colour)

File: challenges/2020/07-handyHaversacks/go/challenge/partTwo.go
package challenge

func countForBag(ruleset map[string]map[string]int, testCl string) int {
	if len(ruleset[testCl]) == 0 {
		return -1
	}

	count := 0

	for childColour := range ruleset[testCl] {
		childBags := countForBag(ruleset, childColour)
		var v int
		if childBags == -1 {
			v = ruleset[testCl][childColour]
		} else {
			v = ruleset[testCl][childColour] * childBags
			v += ruleset[testCl][childColour]
		}
		count += v
	}

	return count

}

func PartTwo(instr string) int {
	return countForBag(parse(instr), targetColour)
}
